DOC-4451 (#76)
couchbaselabs/docs-couchbase-lite: modules/ROOT/pages/_partials/replication-filters.adoc

Replication Filters allow you to have quick control over which documents are stored as the result of a push and/or pull replication.

> Examples of push or pull filters include:

Reviewer: ```suggestion
```

Personal opinion: This sentence doesn't seem necessary.

==== Push Filter

A push filter allows an app to push a subset of a database to the server, which can be very useful in some circumstances.

> For instance, high-priority docs could be pushed first, or docs in a "draft" state could be skipped.

Reviewer: ```suggestion
For instance, high-priority documents could be pushed first, or documents in a "draft" state could be skipped.
```

We probably shouldn't ever abbreviate the word 'documents'. It looks like you don't abbreviate it anywhere else on the page, so using 'documents' is more consistent anyway.

The following example filters out documents whose `type` property is equal to `draft`.

[source]
----
include::{snippet}[tag=replication-push-filter,indent=0]
----

> <1> The callback should follow the semantics of a https://en.wikipedia.org/wiki/Pure_function[pure function].

Reviewer: ```suggestion
<1> The callback should follow the semantics of a https://en.wikipedia.org/wiki/Pure_function[pure function^].
```

Any link that takes you outside of the docs should open in a new tab.

> Otherwise, long running functions would considerably slow down the replicator.

Reviewer: ```suggestion
Otherwise, long running functions would slow down the replicator considerably.
```

Personal opinion: I feel like this reads better.

Furthermore, your callback should not make assumptions about what thread it is being called on.

==== Pull Filter

A pull filter gives an app the ability to validate documents being pulled, and skip ones that fail.
This is an important security mechanism in a peer-to-peer topology with peers that are not fully trusted.

NOTE: Pull replication filters are not a substitute for channels.

> Sync Gateway xref:sync-gateway::data-routing.adoc[channels] are designed to be scalable (documents are filtered on the server) where as a pull replication filter is applied to a document once it has been downloaded.

Reviewer: ```suggestion
Sync Gateway xref:sync-gateway::data-routing.adoc[channels] are designed to be scalable (documents are filtered on the server) whereas a pull replication filter is applied to a document once it has been downloaded.
```

[source]
----

> include::{snippet}[tag=replication-pull-filter,indent=0]

Reviewer: The `<1>` callout is not present within the `replication-pull-filter` region of the snippets for Java, C#, and Objective-C. (It only appears in the snippet for Swift.) Therefore the callout seems orphaned beneath the code blocks.

Reviewer: Which makes me think that the AsciiDoctor [documentation for callouts](https://asciidoctor.org/docs/user-manual/#callouts) is misleading or even outright wrong:

> Each callout number used in a verbatim block **must appear twice**. The first use, which goes within the verbatim block, marks the line being annotated (i.e., the target). The second use, which goes below the verbatim block, defines the annotation text.

Based on this definition, I would assume that a callout would not render unless it is specified both inside and outside of the verbatim block.

Author: Done in https://github.com/couchbase/docs-couchbase-lite/pull/76/commits/828d191a6d291c341b6ca93813feba2e94d983b1

----

> <1> The callback should follow the semantics of a https://en.wikipedia.org/wiki/Pure_function[pure function].

Reviewer: ```suggestion
<1> The callback should follow the semantics of a https://en.wikipedia.org/wiki/Pure_function[pure function^].
```

Any link that takes you outside of the docs should open in a new tab.

> Otherwise, long running functions would considerably slow down the replicator.

Reviewer: ```suggestion
Otherwise, long running functions would slow down the replicator considerably.
```
Personal opinion: I feel like this reads better.
